File: wbox/wbox.py
import subprocess

from wbox.bindings import *


__author__ = 'Tudor'


def run_wbox(cmdline, time_limit_ms, memory_limit_kb):
    # Time limit in 100ns ticks
    time_limit_ticks = LARGE_INTEGER(time_limit_ms * 1000000 / 100)
    working_set_size = memory_limit_kb

    limits = JOBOBJECT_EXTENDED_LIMIT_INFORMATION()
    limits.JobMemoryLimit = working_set_size
    limits.BasicLimitInformation.PerJobUserTimeLimit = time_limit_ticks
    limits.BasicLimitInformation.LimitFlags = \
        JOB_OBJECT_LIMIT_ACTIVE_PROCESS | JOB_OBJECT_LIMIT_JOB_MEMORY | JOB_OBJECT_LIMIT_JOB_TIME
    limits.BasicLimitInformation.ActiveProcessLimit = 1
    limits.BasicLimitInformation.PriorityClass = NORMAL_PRIORITY_CLASS

    job = kernel32.CreateJobObjectA(None, None)
    if not job:
        raise AssertionError('CreateJobObjectA %d' % GetLastError())

    if not kernel32.SetInformationJobObject(job,
                                            JobObjectExtendedLimitInformation,
                                            byref(limits),
                                            sizeof(JOBOBJECT_EXTENDED_LIMIT_INFORMATION)):
        raise AssertionError('SetInformationJobObject %d' % GetLastError())

    siProcess = STARTUPINFO()
    siProcess.cb = sizeof(STARTUPINFO)
    siProcess.dwFlags = STARTF_USESTDHANDLES | STARTF_USESHOWWINDOW
    siProcess.hStdInput = kernel32.GetStdHandle(STD_INPUT_HANDLE)
    siProcess.hStdOutput = kernel32.GetStdHandle(STD_OUTPUT_HANDLE)
    siProcess.hStdError = kernel32.GetStdHandle(STD_ERROR_HANDLE)
    siProcess.wShowWindow = SW_HIDE

    uiJudge = USER_INFO_1()
    uiJudge.usri1_name = c_wchar_p("judge_user")
    uiJudge.usri1_password = c_wchar_p("p455w0rd")
    uiJudge.usri1_priv = 1  # USER_PRIV_USER
    uiJudge.usri1_flags = UF_PASSWD_CANT_CHANGE | UF_SCRIPT | UF_DONT_EXPIRE_PASSWD

    net_err = netapi32.NetUserAdd(NULL, 1, byref(uiJudge), NULL)
    if net_err:
        raise AssertionError('NetUserAdd %d' % net_err)

    piProcess = PROCESS_INFORMATION()
    piProcess.hProcess = INVALID_HANDLE_VALUE
    piProcess.hThread = INVALID_HANDLE_VALUE
    piProcess.dwProcessId = 0
    piProcess.dwThreadId = 0

    if not advapi32.CreateProcessWithLogonW(uiJudge.usri1_name,
                                            c_wchar_p('localhost'),
                                            uiJudge.usri1_password,
                                            0,
                                            NULL,
                                            create_unicode_buffer(subprocess.list2cmdline(cmdline)),
                                            NORMAL_PRIORITY_CLASS | CREATE_SUSPENDED | CREATE_BREAKAWAY_FROM_JOB,
                                            NULL,
                                            NULL,
                                            byref(siProcess),
                                            byref(piProcess)):
        raise AssertionError('CreateProcess %d' % GetLastError())

    net_err = netapi32.NetUserDel(NULL, uiJudge.usri1_name)
    if net_err:
        raise AssertionError('NetUserDel %d' % net_err)

    if kernel32.AssignProcessToJobObject(job, piProcess.hProcess) == 0:
        raise AssertionError('AssignProcessToJobObject %d' % GetLastError())

    if kernel32.ResumeThread(piProcess.hThread) == -1:
        raise AssertionError('ResumeThread %d' % GetLastError())

    if kernel32.CloseHandle(piProcess.hThread) == 0:
        raise AssertionError('CloseHandle(process thread) %d' % GetLastError())

    wait = kernel32.WaitForSingleObject(piProcess.hProcess, time_limit_ms)
    tle = bool(wait)
    if tle:
        kernel32.TerminateProcess(piProcess.hProcess, 0xdeadbeef)
        print 'Process is deadbeef'
    kernel32.WaitForSingleObject(piProcess.hProcess, INFINITE)

    if kernel32.QueryInformationJobObject(job,
                                          JobObjectExtendedLimitInformation,
                                          byref(limits),
                                          sizeof(JOBOBJECT_EXTENDED_LIMIT_INFORMATION),
                                          NULL) == 0:
        raise AssertionError('QueryInformationJobObject %d' % GetLastError())

    err = DWORD()
    kernel32.GetExitCodeProcess(piProcess.hProcess, byref(err))
    print "Process exited with", err.value

    print "Peak mem: %.2fmb (%lu bytes)\n" % (limits.PeakJobMemoryUsed / 1024.0 / 1024, limits.PeakJobMemoryUsed)


test = '''
import urllib2
import os
import getpass

print "Starting on user %s..." % getpass.getuser()

def test(name, func, err):
    print "Testing %s..." % name
    try:
        func()
        print "\tFailed"
    except:
        print "\tSuccess, blocked"


def fopen():
    open(r"C:\file.txt", 'w')


def network():
    urllib2.urlopen('http://dmoj.ca').read()


test('network access', network, 500)
test('file access', fopen, 100)
'''

run_wbox([r'C:\Python27\python.exe', '-c', test], 5000, 65536 * 1024)

File: wbox/wbox.cpp
#define WINVER 0x0502
#define WIN32_LEAN_AND_MEAN

#ifndef UNICODE
#define UNICODE
#endif // UNICODE

#include <windows.h>
#include <lmcons.h>
#include <lmaccess.h>
#include <lmerr.h>
#include <lmapibuf.h>
#include <stdio.h>
#include <stdlib.h>

#include <windows.h>
#include <netfw.h>
#include <objbase.h>
#include <oleauto.h>
#include <stdio.h>

#include <iostream>
#include <windows.h>
#include <windowsx.h>
#include <psapi.h>
#include <tchar.h>
#include <stdio.h>

#pragma comment(lib, "netapi32.lib")
#pragma comment(lib, "ole32.lib")
#pragma comment(lib, "oleaut32.lib")

#include <lm.h>
#include <lmaccess.h>
#include <daogetrw.h>

using namespace std;

#define FAIL(x) return printf("%s failed with %lu\n", x, GetLastError()), -1;

HRESULT WFCOMInitialize(INetFwPolicy2** ppNetFwPolicy2)
{
    HRESULT hr = S_OK;

    hr = CoCreateInstance(
        __uuidof(NetFwPolicy2),
        NULL,
        CLSCTX_INPROC_SERVER,
        __uuidof(INetFwPolicy2),
        (void**)ppNetFwPolicy2);

    if (FAILED(hr))
    {
        printf("CoCreateInstance for INetFwPolicy2 failed: 0x%08lx\n", hr);
        goto error;
    }

error:
    return hr;
}

int main() {
	HRESULT hrComInit = S_OK;
    HRESULT hr = S_OK;

    INetFwPolicy2 *pNetFwPolicy2 = NULL;
    INetFwRules *pFwRules = NULL;
    INetFwRule *pFwRule = NULL;

    long CurrentProfilesBitMask = 0;

    BSTR bstrRuleName = SysAllocString(L"DMOJ Submission Network Block");
    BSTR bstrRuleDescription = SysAllocString(L"Block DMOJ submissions from accessing the network");
    BSTR bstrRuleGroup = SysAllocString(L"DMOJ");
    BSTR bstrRuleApplication = SysAllocString(L"C:\\Python27\\python.exe");
    BSTR bstrRuleLPorts = SysAllocString(L"all");

    hrComInit = CoInitializeEx(
                    0,
                    COINIT_APARTMENTTHREADED
                    );

    if (hrComInit != RPC_E_CHANGED_MODE)
        if (FAILED(hrComInit)) FAIL("CoInitializeEx")

    hr = WFCOMInitialize(&pNetFwPolicy2);
    hr = pNetFwPolicy2->get_Rules(&pFwRules);
    hr = pNetFwPolicy2->get_CurrentProfileTypes(&CurrentProfilesBitMask);

    if ((CurrentProfilesBitMask & NET_FW_PROFILE2_PUBLIC) &&
        (CurrentProfilesBitMask != NET_FW_PROFILE2_PUBLIC))
    {
        CurrentProfilesBitMask &= ~NET_FW_PROFILE2_PUBLIC;
    }

    hr = CoCreateInstance(
                __uuidof(NetFwRule),
                NULL,
                CLSCTX_INPROC_SERVER,
                __uuidof(INetFwRule),
                (void**)&pFwRule);

    pFwRule->put_Name(bstrRuleName);
    pFwRule->put_Description(bstrRuleDescription);
    pFwRule->put_ApplicationName(bstrRuleApplication);
    pFwRule->put_Protocol(NET_FW_IP_PROTOCOL_TCP);
    pFwRule->put_LocalPorts(bstrRuleLPorts);
    pFwRule->put_Direction(NET_FW_RULE_DIR_OUT);
    pFwRule->put_Grouping(bstrRuleGroup);
    pFwRule->put_Profiles(CurrentProfilesBitMask);
    pFwRule->put_Action(NET_FW_ACTION_BLOCK);
    pFwRule->put_Enabled(VARIANT_TRUE);

    hr = pFwRules->Add(pFwRule);

    BOOL bError;
    DWORD milliseconds = 3000;
    size_t mxWorkingSetSize = 65536 * 1024;

    LARGE_INTEGER ticks;
    ticks.QuadPart = milliseconds * 1000000 / 100;

    JOBOBJECT_EXTENDED_LIMIT_INFORMATION extLimits;
    ZeroMemory(&extLimits, sizeof extLimits);
    extLimits.JobMemoryLimit = mxWorkingSetSize;
    extLimits.BasicLimitInformation.PerJobUserTimeLimit = ticks;
    extLimits.BasicLimitInformation.LimitFlags =
            JOB_OBJECT_LIMIT_ACTIVE_PROCESS |
                    JOB_OBJECT_LIMIT_JOB_MEMORY |
                    JOB_OBJECT_LIMIT_JOB_TIME;
    extLimits.BasicLimitInformation.ActiveProcessLimit = 1;
    extLimits.BasicLimitInformation.PriorityClass = NORMAL_PRIORITY_CLASS;

    HANDLE hJobObject;
    hJobObject = CreateJobObject(NULL, NULL);

    bError = SetInformationJobObject(hJobObject, JobObjectExtendedLimitInformation, &extLimits, sizeof extLimits);
    if (!bError) FAIL("SetInformationJobObject");

    STARTUPINFO siProcess;
    ZeroMemory(&siProcess, sizeof siProcess);
    siProcess.cb = sizeof siProcess;

    siProcess.hStdError = GetStdHandle(STD_ERROR_HANDLE);
    siProcess.hStdOutput = GetStdHandle(STD_OUTPUT_HANDLE);
    siProcess.hStdInput = NULL;
    siProcess.dwFlags |= STARTF_USESTDHANDLES;

    PROCESS_INFORMATION piProcess;
    ZeroMemory(&piProcess, sizeof piProcess);

    USER_INFO_1 uiJudge;
    ZeroMemory(&uiJudge, sizeof uiJudge);
    uiJudge.usri1_name = (LPWSTR) L"judge_user";
    uiJudge.usri1_password = (LPWSTR) L"p455w0rd";
    uiJudge.usri1_priv = USER_PRIV_USER;
    uiJudge.usri1_flags = UF_PASSWD_CANT_CHANGE | UF_SCRIPT | UF_DONT_EXPIRE_PASSWD;

    NET_API_STATUS dwUserError = NetUserAdd(NULL, 1, (LPBYTE) &uiJudge, NULL);
    if (dwUserError != NERR_Success) FAIL("NetUserAdd");

    WCHAR wcCmdline[256] = L"C:\\Python27\\python.exe \"C:\\test.py\"";

	int a = TOKEN_QUERY;

    bError = CreateProcessWithLogonW(
            uiJudge.usri1_name,
            L"localhost",
            uiJudge.usri1_password,
            0,
            NULL,
            wcCmdline,
            NORMAL_PRIORITY_CLASS | CREATE_SUSPENDED | CREATE_BREAKAWAY_FROM_JOB ,
            NULL,
            NULL,
            &siProcess,
            &piProcess);

    if (!bError) FAIL("CreateProcess");

    dwUserError = NetUserDel(NULL, uiJudge.usri1_name);
    if (dwUserError != NERR_Success) FAIL("NetUserDel");

    HANDLE hProcess = piProcess.hProcess;

    bError = AssignProcessToJobObject(hJobObject, hProcess);
    if (!bError) FAIL("AssignProcessToJobObject");

    ResumeThread(piProcess.hThread);
    bError = CloseHandle(piProcess.hThread);
    if (!bError) FAIL("CloseHandle(process thread)");

    DWORD dwWait = WaitForSingleObject(piProcess.hProcess, milliseconds);
    BOOL bTle = dwWait != WAIT_OBJECT_0;
    if (bTle)
        TerminateProcess(piProcess.hProcess, 0xdeadbeef);
    WaitForSingleObject(piProcess.hProcess, INFINITE);
    DWORD dwErr;
    GetExitCodeProcess(piProcess.hProcess, &dwErr);
    CloseHandle(piProcess.hProcess);

    bError = QueryInformationJobObject(hJobObject, JobObjectExtendedLimitInformation, &extLimits, sizeof extLimits, NULL);
    if (!bError) FAIL("QueryInformationJobObject");

    printf("---\n");
    printf("Exit code: %lu\n", dwErr);
    printf("TLE: %d\n", bTle);
    printf("Peak mem: %.2fmb (%lu bytes)\n", extLimits.PeakJobMemoryUsed / 1024.0 / 1024, extLimits.PeakJobMemoryUsed);

    hr = pFwRules->Remove(bstrRuleName);

    return 0;
}
